runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/External.java

```diff
@@ -263,9 +268,11 @@ public OutputT expand(InputT input) {
       return toOutputCollection(outputMapBuilder.build());
     }
 
-    private RunnerApi.Components resolveArtifacts(RunnerApi.Components components) {
-      if (components.getEnvironmentsMap().values().stream()
-          .allMatch(env -> env.getDependenciesCount() == 0)) {
+    private RunnerApi.Components resolveArtifacts(
+        RunnerApi.Components components, Set<String> originalEnvIds) {
```

Comment:
It would be a cleaner API to pass the set of environments that should be expanded than pass in the set that should not be. (Even better might be to simply take a Map<string, Environment> rather than the whole surrounding Components object.)

Reply:
Done.